### PR TITLE
[feat] Add new `ci_extras` test variable that allows passing test-specific options to the pipeline generated by `--ci-generate`

### PR DESCRIPTION
**Pull request by donion74**
This PR allows to use Gitlab CI options to filter tests, e.g. based on [downstream pipelines](https://docs.gitlab.com/ee/ci/pipelines/downstream_pipelines.html) or [changes of the file](https://docs.gitlab.com/ee/ci/jobs/job_control.html). A minimal working example is provided in the repository https://gitlab.com/latticeqcd/reframe-debug, see [hello1.py](https://gitlab.com/latticeqcd/reframe-debug/-/blob/master/hello1.py)
- [Pipeline 810057836](https://gitlab.com/latticeqcd/reframe-debug/-/pipelines/810057836) runs for a branch that is in merge request, but has unmodified source file. Hence the test `HelloMergeRequest` is run.
- [Pipeline 810072116](https://gitlab.com/latticeqcd/reframe-debug/-/pipelines/810072116) now has modified source file and thus runs additionally `HelloModified`.

### CODE REVIEW

**Comment by donion74** [comment]
@jenkins-cscs Yes, I just updated the description.

**Review by vkarak — CHANGES_REQUESTED** [review]
Hi @donion74, thanks for the  PR!

I like the idea as it's simple, yet powerful. I think it essentially addresses what #2420 tries to do. @mahendrapaipuri Wdyt? Would that feature work for your case as well?

Other than that we'd need a unit test or expand the existing one in `test_ci.py` that basically exercises also this feature and a bit of fine tuning of the docs as this is not about adding "options" to the CI: users can essentially extend per test basis the generated child pipeline.

**Comment by mahendrapaipuri** [comment]
@donion74 @vkarak I think it should work in my case as well. From what I see we can arbitrary configure the child pipeline within each "test", which was exactly what we were looking for. I really like this PR as it is a very simple and efficient solution.

@donion74 What happens if we want to configure directives that are explicitly generated by current CI code? For instance, ReFrame already emits [artifacts](https://github.com/reframe-hpc/reframe/blob/9de010f946e49546e4c34e6b08349ffdef03b98d/reframe/frontend/ci.py#L68-L69) to include the report file. If user wants to add more artifacts to it, with this PR, the user can set `ci_options` as

```
ci_options = {"artifacts": {"paths": ['/path1/to/save', '/path2/to/save']}}
```

which will end up generating a YAML file as follows in your example:

```
HelloModified:
  stage: rfm-stage-0
  script:
  - reframe --prefix=rfm-stage/${CI_COMMIT_SHORT_SHA}  -c /builds/latticeqcd/reframe-debug  --report-file=HelloModified-report.json  --report-junit=HelloModified-report.xml  -n '^HelloModified$' -r
  artifacts:
    paths:
    - HelloModified-report.json
  needs: []
  artifacts:
   paths:
     - /path1/to/save
     - /path2/to/save
```
So, the question is if the GitLab CI parser (in future if other CI engines will be supported, their parsers as well) is smart enough to merge them? Could you maybe do a quick test?

File: reframe/core/pipeline.py [hunk 1]
```diff
@@ -963,6 +963,20 @@ def pipeline_hooks(cls):
     #: :default: :class:`True`
     build_locally = variable(typ.Bool, value=True, loggable=True)
 
+    #: .. versionadded:: ??
```
[the remark this answers is not in the document]
Comment (donion74) [reply reframe/core/pipeline.py]: Added it.

File: reframe/core/pipeline.py [hunk 2]
```diff
@@ -963,6 +963,20 @@ def pipeline_hooks(cls):
...
+    #:
+    #: Example: If we want a pipeline to run only when files in
+    #: backend or src/main.c have changed, we add:
+    #:
+    #: ci_options = {"only": {"changes": ["backend/*", "src/main.c"]}
+    #:
+    #: :type: `dict`
+    #: :default: ``{}``
+    ci_options = variable(typ.Dict[str, object],
+                          value={}, loggable=True)
```
[the remark this answers is not in the document]
Comment (donion74) [reply reframe/core/pipeline.py]: - removed loggable option.
- renamed it `ci_extras`.
- added the `gitlab` key.

**Comment by donion74** [comment]
@mahendrapaipuri Good point. If I try to reproduce your suggestions, the `artifacts` key is overwritten:

```
    ci_extras['gitlab'] = {**ci_extras['gitlab'], 
        "rules": [{"changes": ["src/hello.c"]}],
        'artifacts': {"paths": ['pipeline.yml', '.gitlab-ci.yml']}}
```
produces this pipeline:

```
HelloModified:
  stage: rfm-stage-0
  script:
  - reframe --prefix=rfm-stage/${CI_COMMIT_SHORT_SHA} -C /home/anian/phd/2-openqxd/tests/config//settings.py -c . -R --report-file=HelloModified-report.json  --report-junit=HelloModified-report.xml  -n '^HelloModified$' -r
  artifacts:
    paths:
    - pipeline.yml
    - .gitlab-ci.yml
  needs: []
  rules:
  - changes:
    - src/hello.c
```

I guess this is related to https://stackoverflow.com/q/14902299, and could be solved by the suggestion https://stackoverflow.com/a/61416215.

**Comment by donion74** [comment]
I updated the example git repo. Here are two pipelines that demonstrate the use of `changes`:
- https://gitlab.com/latticeqcd/reframe-debug/-/pipelines/821682420
- https://gitlab.com/latticeqcd/reframe-debug/-/pipelines/821685043

**Comment by mahendrapaipuri** [comment]
Thanks @donion74 for the tests. So, if Gitlab CI is not merging them, we need to do it within ReFrame. So, currently ReFrame is emitting pipelines using directives `stage`, `script`, `artifacts` and `needs`.

- `stage` and `needs` are important for dependency resolution. So, if user defined `ci_extras` have these directives we need to ignore them 
- Document clearly that users need to use `before_script` and `after_script` for executing auxillary commands within a job. If user still passes `script` directive within `ci_extras`, we need to ignore it as we do not know whether to execute them before or after ReFrame job. A warning would be nice to have when this case arises.
- `artifacts` need to be merged.

I think this should do nicely to configure CI pipelines. What do you think @vkarak?

**Comment by vkarak** [comment]
@mahendrapaipuri I agree with your suggestion. For `needs`, `stage` and `script` we need to issue an error I think and if the user passes `script` specifically, we should ask her to set either the `before_script` or `after_script` instead. `artifacts` should indeed be merged, but we should not bother implementing a generic JSON object merger but rather a specific one for the artifacts; reframe knows exactly what it emits for artifacts, so it can combine it with the user's artifacts accordingly.

**Review by vkarak — APPROVED** [review]
I updated the PR implementing the restrictions we just discussed and add also unit tests. Now lgtm. @donion74 @mahendrapaipuri If you confirm that it works for your use cases, then we can merge this PR.

**Comment by mahendrapaipuri** [comment]
@vkarak Cheers @vkarak!! Yes, this would work for my cases and this PR is quite handy. Cheers @donion74 !!

**Comment by donion74** [comment]
Thanks for implementing the changes @vkarak , my pipeline still works as intended: https://gitlab.com/latticeqcd/reframe-debug/-/pipelines/827286511

The only downside I notice is that subsequent tests overwrite the variable `ci_extras`:
See the test [hello1.py](https://gitlab.com/latticeqcd/reframe-debug/-/blob/example/hello1.py), and the corresponding [pipeline](https://gitlab.com/latticeqcd/reframe-debug/-/pipelines/827454922), where I print the content of `ci_extras` in [line 43](https://gitlab.com/latticeqcd/reframe-debug/-/jobs/4057630126) and [line 43](https://gitlab.com/latticeqcd/reframe-debug/-/jobs/4057630127).

**Comment by mahendrapaipuri** [comment]
@vkarak I managed to find the problem but not sure about the solution. Lets take a simplified test case:

```
@rfm.simple_test
class HelloArtifacts(rfm.RegressionTest):
    """only runs in merge requests and stores additional artifacts"""
    valid_systems = ['*']
    valid_prog_environs = ['*']
    sourcepath = 'hello.c'
    ci_extras['gitlab'] = {"rules": [{"if": "$CI_MERGE_REQUEST_ID"}],
                           'artifacts': {"paths": ['.gitlab-ci.yml']}}

    @sanity_function
    def assert_hello(self):
        print(self.ci_extras['gitlab'])
        return sn.assert_found(r'Hello, World\!', self.stdout)

@rfm.simple_test
class Simple(rfm.RegressionTest):
    """always runs"""
    valid_systems = ['*']
    valid_prog_environs = ['*']
    sourcepath = 'hello.c'
    desc = 'Simple'

    @sanity_function
    def assert_true(self):
        print(self.ci_extras)
        return True
```

This would produce a `pipeline.yml` as follows:

```
cache:
  key: ${CI_COMMIT_REF_SLUG}
  paths:
  - rfm-stage/${CI_COMMIT_SHORT_SHA}
stages:
- rfm-stage-0
Simple:
  stage: rfm-stage-0
  script:
  - reframe --prefix=rfm-stage/${CI_COMMIT_SHORT_SHA}  -c .  --report-file=Simple-report.json  --report-junit=Simple-report.xml -vvvvv -n '^Simple$' -r
  artifacts:
    paths:
    - Simple-report.json
    - .gitlab-ci.yml
  needs: []
  rules:
  - if: $CI_MERGE_REQUEST_ID
HelloArtifacts:
  stage: rfm-stage-0
  script:
  - reframe --prefix=rfm-stage/${CI_COMMIT_SHORT_SHA}  -c .  --report-file=HelloArtifacts-report.json  --report-junit=HelloArtifacts-report.xml -vvvvv -n '^HelloArtifacts$' -r
  artifacts:
    paths:
    - HelloArtifacts-report.json
    - .gitlab-ci.yml
  needs: []
  rules:
  - if: $CI_MERGE_REQUEST_ID
```
This is the buggy behaviour where `ci_extras` in `Simple` test are being populated from `HelloArtifacts`. However, if we change the way we set `ci_extras` to

```
 ci_extras = {''gitlab'': {"rules": [{"if": "$CI_MERGE_REQUEST_ID"}],
                           'artifacts': {"paths": ['.gitlab-ci.yml']}}}
```
we get correct behaviour giving us following `pipeline.yml`

```
cache:
  key: ${CI_COMMIT_REF_SLUG}
  paths:
  - rfm-stage/${CI_COMMIT_SHORT_SHA}
stages:
- rfm-stage-0
Simple:
  stage: rfm-stage-0
  script:
  - reframe --prefix=rfm-stage/${CI_COMMIT_SHORT_SHA}  -c .  --report-file=Simple-report.json  --report-junit=Simple-report.xml -vvvvv -n '^Simple$' -r
  artifacts:
    paths:
    - Simple-report.json
  needs: []
HelloArtifacts:
  stage: rfm-stage-0
  script:
  - reframe --prefix=rfm-stage/${CI_COMMIT_SHORT_SHA}  -c .  --report-file=HelloArtifacts-report.json  --report-junit=HelloArtifacts-report.xml -vvvvv -n '^HelloArtifacts$' -r
  artifacts:
    paths:
    - HelloArtifacts-report.json
    - .gitlab-ci.yml
  needs: []
  rules:
  - if: $CI_MERGE_REQUEST_ID
```
@vkarak , I hope that can give you an idea where to look at!!

File: reframe/core/pipeline.py [hunk 3]
```diff
@@ -963,6 +963,19 @@ def pipeline_hooks(cls):
...
     build_locally = variable(typ.Bool, value=True, loggable=True)
 
+    #: .. versionadded:: 4.2
+    #:
+    #: Options for CI pipeline passed in JSON format
+    #:
+    #: Example: If we want a pipeline to run only when files in
+    #: backend or src/main.c have changed, we add:
+    #:
+    #: ci_options = {"only": {"changes": ["backend/*", "src/main.c"]}
```
Comment (mahendrapaipuri) [inline reframe/core/pipeline.py:973]: In the example, it has to be `ci_extras` and dict must have `gitlab` key

**Comment by vkarak** [comment]
I accidentally deleted my previous comment. If you instead set the `ci_extras` completely and not just the `'gitlab'` element, it should work. I am still investigating why this is happening and whether it is a corner case bug of the variables mechanism. I have fixed the PR to avoid this from happening by setting the default to `None`.

**Comment by vkarak** [comment]
It is indeed a bug introduced in 4.0. See #2848. I will revert this PR's implementation using `{}` as the default value and @donion74 meanwhile update your tests to set the `ci_extras` as a whole until this bug is fixed.

**Comment by mahendrapaipuri** [comment]
@vkarak @donion74 Seems like this regression is coming from this [commit](https://github.com/reframe-hpc/reframe/commit/d01138163e6ab5093827d9807f890435ad629b26). In the CI tests, we are explicitly setting the `--report-file` with relative path. The [basedir](https://github.com/reframe-hpc/reframe/blob/6e22addc096cb61640baaa8d9b8851c4975beda3/reframe/frontend/cli.py#L1339-L1344) is being calculated on relative path and wrapping it inside `os.path.abspath` should fix the problem.

**Comment by vkarak** [comment]
@mahendrapaipuri Good catch, would you mind opening an issue? The `--report-file` should work with relative paths...

**Comment by mahendrapaipuri** [comment]
@vkarak Issue #2849

**Comment by mahendrapaipuri** [comment]
@vkarak @donion74 I was meaning to ask you if it is worth to include this new `ci_extras` in the ReFrame config file. So, instead of setting a default value within the class, we get it from the config file. 

The use case is that users can define a global CI config in the ReFrame config and override it for specific tests within the test. This is especially useful for users who wish to use same CI config for all tests as they will not need to set the config for "each" test.

**Comment by donion74** [comment]
This would be very useful. The initial reason I came up with that was to run the Gitlab CI only in merge request pipelines, so I would need to add the same 
```
  rules:
  - if: $CI_MERGE_REQUEST_ID
 ```
in all the tests. Your idea would avoid duplicate code.

**Comment by vkarak** [comment]
@mahendrapaipuri I suggest opening an issue about that. There is also an old one that requests essentially this for any test variable #1500.

**Pull request merged**